src/deriveNativeScriptHash.c

```diff
@@ -0,0 +1,520 @@
+
+	ctx->complexScripts[ctx->level].ui_type = UI_SCRIPT_N_OF_K;
+	UI_DISPLAY_SCRIPT(UI_SCRIPT_N_OF_K);
+}
+
+static void deriveNativeScriptHash_handleComplexScriptStart(read_view_t* view)
+{
+	TRACE_BUFFER(view->begin, (size_t) (view->end - view->begin));
+
+	expectsMoreScripts();
+
+	// check if we can increase the level without breaking the MAX_SCRIPT_DEPTH constraint
+	VALIDATE(ctx->level + 1 < MAX_SCRIPT_DEPTH, ERR_DATA_TOO_LARGE);
+	ctx->level++;
+	explicit_bzero(&ctx->complexScripts[ctx->level], SIZEOF(ctx->complexScripts[ctx->level]));
+
+	uint8_t nativeScriptType = parse_u1be(view);
+	TRACE("native complex script type = %u", nativeScriptType);
+
+	switch(nativeScriptType) {
+#	define  CASE(TYPE, HANDLER) case TYPE:\
+		ctx->complexScripts[ctx->level].remainingScripts = parse_u4be(view);\
+		ctx->complexScripts[ctx->level].totalScripts = ctx->complexScripts[ctx->level].remainingScripts;\
+		HANDLER(view);\
+		break;
+		CASE(NATIVE_SCRIPT_ALL, deriveNativeScriptHash_handleAll);
+		CASE(NATIVE_SCRIPT_ANY, deriveNativeScriptHash_handleAny);
+		CASE(NATIVE_SCRIPT_N_OF_K, deriveNativeScriptHash_handleNofK);
+#	undef   CASE
+	default:
+		THROW(ERR_INVALID_DATA);
+	}
+}
+
+// Simple native scripts
+
+static void deriveNativeScriptHash_handleDeviceOwnedPubkey(read_view_t* view)
+{
+	view_skipBytes(view, bip44_parseFromWire(&ctx->scriptContent.pubkeyPath, VIEW_REMAINING_TO_TUPLE_BUF_SIZE(view)));
+
+	VALIDATE(view_remainingSize(view) == 0, ERR_INVALID_DATA);
+	TRACE("pubkey given by path:");
+	BIP44_PRINTF(&ctx->scriptContent.pubkeyPath);
+	PRINTF("\n");
+
+	uint8_t pubkeyHash[ADDRESS_KEY_HASH_LENGTH];
+	bip44_pathToKeyHash(&ctx->scriptContent.pubkeyPath, pubkeyHash, ADDRESS_KEY_HASH_LENGTH);
+	nativeScriptHashBuilder_addScript_pubkey(&ctx->hashBuilder, pubkeyHash);
+
+	UI_DISPLAY_SCRIPT(UI_SCRIPT_PUBKEY_PATH);
+}
+
+static void deriveNativeScriptHash_handleThirdPartyPubkey(read_view_t* view)
+{
+	VALIDATE(view_remainingSize(view) == ADDRESS_KEY_HASH_LENGTH, ERR_INVALID_DATA);
+	view_memmove(ctx->scriptContent.pubkeyHash, view, ADDRESS_KEY_HASH_LENGTH);
+
+	TRACE_BUFFER(ctx->scriptContent.pubkeyHash, ADDRESS_KEY_HASH_LENGTH);
+
+	nativeScriptHashBuilder_addScript_pubkey(&ctx->hashBuilder, ctx->scriptContent.pubkeyHash);
+
+	UI_DISPLAY_SCRIPT(UI_SCRIPT_PUBKEY_HASH);
+}
+
+static void deriveNativeScriptHash_handlePubkey(read_view_t* view)
+{
+	uint8_t pubkeyType = parse_u1be(view);
+	TRACE("pubkey type = %u", pubkeyType);
+
+	switch(pubkeyType) {
+	case KEY_REFERENCE_PATH:
+		deriveNativeScriptHash_handleDeviceOwnedPubkey(view);
+		break;
+	case KEY_REFERENCE_HASH:
+		deriveNativeScriptHash_handleThirdPartyPubkey(view);
+		break;
+	// any other value for the pubkey type is invalid
+	default:
+		THROW(ERR_INVALID_DATA);
+	}
+}
+
+// TRACE doesn't handle uint64_t so we split it into two uint32_t
+#define TRACE_UINT64(varname, n) TRACE(varname "= 0x%x%x", (uint32_t) (n >> 32), (uint32_t) n)
```

Comment:
Ah, now I see, it's not a macro, it's a function called `str_traceUint64`, which honestly seems much more complicated then it's needed for tracing a uint64. Do you want me to change it?
In the function call the uint64 is first converted to a string and then traced. My macro is in my opinion much more simpler for debug purposes.

Reply:
I think we should do the same thing everywhere when we log uint64. (You are welcome to propose improvements to the existing code, e.g. adding a variable name to TRACE_UINT64.)

One purpose of tracing is that you can compare the values from usbtool to those logged in javascript (I use plain `console.log` since I don't know anything better...). Does your macro makes the comparison straightforward? If not, I would be wary of using it as a permanent solution.

Reply:
It probably doesn't unfortunately, it displays the uint64 in it's hex form, and if I am correctly aware the `console.log` displays is as a normal number, so I will probably switch to the already existing solution